### PR TITLE
ls: improvements on time handling

### PR DESCRIPTION
**Pull request by LevitatingBusinessMan**
I added some extra compatibility to `ls`.
Creation/Birth times.
And time styles.

Still needs work.
`FORMAT+`, `posix-` are not yet implemented as styles.

I moved `ls` from an old version of `time` to a new version of chrono which supports translating locales.
To do this I had to update `uucore` to `time 0.1.43` to match the time package used in the new chrono release.
Sadly I haven't found a good way yet to find a users locale so the translating isn't implemented yet.

Some tests are still need to be made.
And I am a bit skeptical of how `--full-time`, `--long` and `--time-style` play together right now, so I shall make sure that's all overriding each other correctly in another commit.

If I add more features or tests I'll update this PR

I hope to cover the whole time part of https://github.com/uutils/coreutils/issues/1872 soon.

Edit: Seems like sometime when switching to `chrono` the `--full-time` arg started giving crashes. Something I'll have to investigate.

### CODE REVIEW

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
I think this is a great effort! Kudos for taking on such a complicated set of options. There are a couple of points where I think it could be improved, mostly minor things though. I am a little bit worried about the references to the GNU source code, see the comment below about that for more info.

File: src/uu/ls/src/ls.rs [hunk 1]
```diff
@@ -1128,18 +1204,41 @@ fn get_system_time(md: &Metadata, config: &Config) -> Option<SystemTime> {
...
+            match config.time_style {
+                TimeStyle::FullIso => time.format("%Y-%m-%d %H:%M:%S.%N %z"),
+                TimeStyle::LongIso => time.format("%Y-%m-%d %H:%M"),
+                TimeStyle::Iso => time.format(if recent {
+                    "%Y-%m-%d %H:%M"
+                } else {
+                    "%Y-%m-%d "
+                }),
+                TimeStyle::Locale => {
+                    //https://github.com/coreutils/coreutils/blob/master/src/ls.c#L759
```
[the remark this answers is not in the document]
Comment (sylvestre) [reply src/uu/ls/src/ls.rs:1230]: yeah

File: tests/by-util/test_ls.rs [hunk 2]
```diff
@@ -567,6 +567,21 @@ fn test_ls_long_ctime() {
     assert!(result.stdout.contains("???"));
 }
 
+#[test]
+fn test_ls_order_birthtime() {
+    let scene = TestScenario::new(util_name!());
+    let at = &scene.fixtures;
+
+    at.touch("test-birthtime-1");
+    at.touch("test-birthtime-2");
```
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply tests/by-util/test_ls.rs:576]: Also, the tests seem to fail, probably because there is almost no time between the creation of these files, so maybe sleeping for 500 milliseconds would help?

File: src/uu/ls/src/ls.rs [hunk 3]
```diff
@@ -1128,18 +1204,41 @@ fn get_system_time(md: &Metadata, config: &Config) -> Option<SystemTime> {
...
+            //According to GNU a Gregorian year has 365.2425 * 24 * 60 * 60 == 31556952 seconds on the average.
+            //https://github.com/coreutils/coreutils/blob/master/src/ls.c#L4385
+            let recent = time + chrono::Duration::seconds(31556952 / 2) > chrono::Local::now();
+
+            //For reference see https://github.com/coreutils/coreutils/blob/master/src/ls.c#L2416
+            match config.time_style {
+                TimeStyle::FullIso => time.format("%Y-%m-%d %H:%M:%S.%N %z"),
+                TimeStyle::LongIso => time.format("%Y-%m-%d %H:%M"),
+                TimeStyle::Iso => time.format(if recent {
+                    "%Y-%m-%d %H:%M"
```
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:1225]: > This is not the format I'm getting on my machine. I get the recent files formatted like this:
> 
> ```
> 11-25 14:54 (%m-%d %H:%M)
> ```

Yep, that's a mistake. 



> Also, we currently don't do anything clever in the long format to ensure everything is lined up, so I think this might mess up the spacing a little bit?

I'll look into padding that.

File: src/uu/ls/src/ls.rs [hunk 4]
```diff
@@ -1128,18 +1204,41 @@ fn get_system_time(md: &Metadata, config: &Config) -> Option<SystemTime> {
...
+                    "%Y-%m-%d "
+                }),
+                TimeStyle::Locale => {
+                    //https://github.com/coreutils/coreutils/blob/master/src/ls.c#L759
+                    let fmt = if recent { "%b %e %H:%M" } else { "%b %e  %Y" };
+
+                    //In this version of chrono translating can be done
+                    //The function is chrono::datetime::DateTime::format_localized
+                    //However it's currently still hard to get the current pure-rust-locale
+                    //So it's not yet implemented
```
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:1236]: > Maybe just leave it out for now then. Makes it more clear that it is missing.

I am not sure exactly _what_ you think should be left out. The comment or the locale setting?

File: tests/by-util/test_ls.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply tests/by-util/test_ls.rs:576]: > there is almost no time between the creation of these files
This was my guess at the problem as well.

File: src/uu/ls/src/ls.rs [hunk 5]
```diff
@@ -779,9 +824,39 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
...
+        )
 
     // Positional arguments
-        .arg(Arg::with_name(options::PATHS).multiple(true).takes_value(true));
+        .arg(Arg::with_name(options::PATHS).multiple(true).takes_value(true))
+
+        .after_help(indoc!(
+        "The TIME_STYLE argument can be full-iso, long-iso, iso.
+        Also the TIME_STYLE environment variable sets the default style to use.
+        "));
```
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:859]: We are already using `unindent` which is part of `indoc` in the main package.

I don't know what exactly you mean by "I think formatting this like the help for the time option"
If FORMAT+ and posix- get added this extra information will get quite a bit larger.
And I also think we should add the color and size information at some point.

[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:859]: Ah I see, that's alright then!

File: src/uu/ls/src/ls.rs [hunk 6]
```diff
@@ -779,9 +824,39 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
...
+                ])
+        )
+        .arg(
+            Arg::with_name(options::FULL_TIME)
+            .long(options::FULL_TIME)
+            .help("like -l --time-style=full-iso")
+            .overrides_with_all(&[
+                options::TIME_STYLE,
+                options::format::LONG,
+            ])
```
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:850]: > I suggest looking at how -o, -g and -n are implemented
I already looked at those exact implementations but I don't think the issue is actually exactly the same for them.

> ls --full-time -l should still display the long iso format
That does work, but for instance ` --full-time --time-style=iso` doesn't display long options right now.

File: src/uu/ls/src/ls.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:1230]: It's just for referencing what options correspond to what strfmt format, but I'll remove them.

File: src/uu/ls/src/ls.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:1236]: Yeah the whole setting, I think it makes sense to just add it back in when it's actually being implemented.

File: src/uu/ls/src/ls.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:1230]: Should I also remove them from the previous commit? Overwrite history?

File: src/uu/ls/src/ls.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:850]: Yeah `clap` can be quite tricky with these sort of multi-functional flag. Could you also check whether
```
ls --full-time -C
ls --full-time -Cl
```
show the columns format and the long format with `full-iso`, respectively?

> That does work, but for instance --full-time --time-style=iso doesn't display long options right now.

This is probably because `--full-time` has `--time-style` as an override, but removing that will probably result in one always being chosen over the other.

If I understand correctly this argument sort of requires you to do the strategy for `-o` twice, once for the format and once for the time-style, which means
- Don't add any overrides in `clap`.
- Get the index of the last "regular" format argument, compare with the last index of `--full-time`. If the of the `--full-time` index is greater set the format to `long`.
- Get the index of the last "regular" time-style argument, compare with the last index of `--full-time`. If the index of `--full-time` is greater, set the time-style to `full-iso`.

That should probably work right? However, I find this method a bit weird, so it would be great if you find a better method.

File: src/uu/ls/src/ls.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:1236]: It's the default setting though. And it's formatting is specific, so it makes sense to leave it in but just not do any translating yet.

[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:1236]: Okay, that sounds fair!

File: src/uu/ls/src/ls.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sylvestre) [reply src/uu/ls/src/ls.rs:1230]: yeah, if you copy GNU/coreutils, please rewrite the functions (sorry)

[the remark this answers is not in the document]
Comment (sylvestre) [reply src/uu/ls/src/ls.rs:1230]: @LevitatingBusinessMan do you have a plan for this?
thanks

[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:1230]: @sylvestre Like I said I only referenced what strfmt should correspond with what option, you can't really say there is any GNU code in my codebase. Anyhow I'll just rewrite it, I just have to find the time to continue work on this branch. School is keeping me busy right now.

[the remark this answers is not in the document]
Comment (sylvestre) [reply src/uu/ls/src/ls.rs:1230]: If this is just a reference, this is fine then.

File: src/uu/ls/src/ls.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:1225]: I don't think the padding is an issue but you could QA that for me if I finish this up

**Comment by LevitatingBusinessMan** [comment]
At this point I only have to write more tests. @tertsdiepraam could you make a list of what kind of tests you'd like to see?

**Comment by tertsdiepraam** [comment]
Looking good! I think we should at least have tests that
- check all the different time formats
- check whether all the options can be overridden correctly, so things like
```bash
ls --time-style=long-iso --time-style=iso
ls --time-style=iso --full-time
ls --full-time --time-style=iso --full-time
```
(I think you got most of this correctly but it's still nice to have some tests)
- check whether `--full-iso` can be overridden by all other format options (especially `-m` and `-x`):
```bash
ls --full-time -x # should be across format
ls --full-time -xl # should be long format with full iso
```
- check `birth-time` in such a way that distinguishes it from other time formats, like I described here https://github.com/uutils/coreutils/pull/1986#discussion_r605417033

That should cover it I think. You can also add anything that you found complicated or weird behaviour from GNU.

Also note that you should probably merge the upstream master to fix some conflicts.

**Comment by tertsdiepraam** [comment]
Maybe for a next PR, I think the locale for the time can be read from the `LC_TIME` environment variable, but I'm not sure how complicated it is to actually implement the different formats. Explained here under `'locale'`: https://www.gnu.org/software/coreutils/manual/html_node/Formatting-file-timestamps.html#Formatting-file-timestamps

File: src/uu/ls/Cargo.toml [hunk 7]
```diff
@@ -15,15 +15,17 @@ edition = "2018"
...
 lazy_static = "1.0.1"
 number_prefix = "0.4"
 term_grid = "0.1.5"
 termsize = "0.1.6"
-time = "0.1.40"
 unicode-width = "0.1.5"
-uucore = { version=">=0.0.7", package="uucore", path="../../uucore", features=["entries", "fs"] }
-uucore_procs = { version=">=0.0.5", package="uucore_procs", path="../../uucore_procs" }
+uucore = { version = ">=0.0.7", package = "uucore", path = "../../uucore", features = ["entries", "fs"] }
+uucore_procs = { version = ">=0.0.5", package = "uucore_procs", path = "../../uucore_procs" }
```
Comment (sylvestre) [inline src/uu/ls/Cargo.toml:28]: please revert this unrelated change. thanks

```suggestion
uucore = { version=">=0.0.7", package="uucore", path="../../uucore", features=["entries", "fs"] }
uucore_procs = { version=">=0.0.5", package="uucore_procs", path="../../uucore_procs" }
```

File: src/uucore/Cargo.toml [hunk 8]
```diff
@@ -24,7 +24,7 @@ thiserror = { version="1.0", optional=true }
 lazy_static = { version="1.3", optional=true }
 nix = { version="<= 0.13", optional=true }
 platform-info = { version="<= 0.1", optional=true }
-time = { version="<= 0.1.42", optional=true }
+time = { version="<= 0.1.43", optional=true }
```
Comment (sylvestre) [inline src/uucore/Cargo.toml:27]: this is a necessary update?

File: tests/by-util/test_ls.rs [hunk 9]
```diff
@@ -567,6 +567,21 @@ fn test_ls_long_ctime() {
...
+    let result = scene.ucmd().arg("--time=birth").arg("-t").run();
+
+    #[cfg(not(windows))]
+    assert_eq!(result.stdout, "test-birthtime-2\ntest-birthtime-1\n");
+    #[cfg(windows)]
+    assert_eq!(result.stdout, "test-birthtime-2  test-birthtime-1\n");
+}
```
[the remark this answers is not in the document]
Comment (sylvestre) [reply tests/by-util/test_ls.rs:583]: ditto

**Review by sylvestre — CHANGES_REQUESTED** [review]
Looks great, some minor needs

File: src/uu/ls/Cargo.toml (hunk 7 above)
Comment (LevitatingBusinessMan) [reply src/uu/ls/Cargo.toml:28]: @sylvestre This is rustfmt formatting the toml.

File: src/uucore/Cargo.toml (hunk 8 above)
Comment (LevitatingBusinessMan) [reply src/uucore/Cargo.toml:27]: @sylvestre I otherwise had issues with using the new `chrono` releases. I mentioned this previously in my PR body text.

File: src/uu/ls/src/ls.rs [hunk 10]
```diff
@@ -1128,18 +1204,41 @@ fn get_system_time(md: &Metadata, config: &Config) -> Option<SystemTime> {
...
+                }),
+                TimeStyle::Locale => {
+                    //https://github.com/coreutils/coreutils/blob/master/src/ls.c#L759
+                    let fmt = if recent { "%b %e %H:%M" } else { "%b %e  %Y" };
+
+                    //In this version of chrono translating can be done
+                    //The function is chrono::datetime::DateTime::format_localized
+                    //However it's currently still hard to get the current pure-rust-locale
+                    //So it's not yet implemented
+
```
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply src/uu/ls/src/ls.rs:1226]: I already did, I run rustfmt automatically whenever I save a file. There is should be no issue here.

**Comment by sylvestre** [comment]
Sorry but could you please fix the conflicts? thanks

**Comment by LevitatingBusinessMan** [comment]
@tertsdiepraam I am not sure how to test the styles, I could compare with a regex but I'd more rather be able to set the modification time of a file to a unix timestamp.

**Comment by tertsdiepraam** [comment]
I haven't found an easy way to set the modification time yet, so regex sounds good to me!

**Comment by LevitatingBusinessMan** [comment]
```
ls --full-time -x # should be across format
ls --full-time -xl # should be long format with full iso
```

@tertsdiepraam This doesn't work yet, and I honestly don't know how to make it work efficiently... `--full-time` is currently dominant over all short format options.

**Comment by tertsdiepraam** [comment]
@LevitatingBusinessMan I think you're very close, because I think you already got it right with time part of `--full-time`. There is already some logic in place for the format here, so I think adding the `--full-time` option to this array (line 283) might do the trick:
```rust
if [
    options::format::LONG_NO_OWNER,
    options::format::LONG_NO_GROUP,
    options::format::LONG_NUMERIC_UID_GID,
]
```
and then remove the `--full-time` option here (line 251):
```rust
} else if options.is_present(options::format::LONG)
    || options.is_present(options::FULL_TIME)
{
```
Does that help?

**Comment by LevitatingBusinessMan** [comment]
Last commit should do it?

**Comment by tertsdiepraam** [comment]
Nice! If those tests indeed pass then that should do it, it looks good to me!

**Comment by sylvestre** [comment]
Sorry but Ubuntu is failing
```

---- test_ls::test_ls_order_birthtime stdout ----
current_directory_resolved: 
touch: /tmp/.tmpiUSBi3/test-birthtime-1
touch: /tmp/.tmpiUSBi3/test-birthtime-2
touch: /tmp/.tmpiUSBi3/test-birthtime-1
run: /target/i686-unknown-linux-musl/debug/coreutils ls --time=birth -t
thread 'test_ls::test_ls_order_birthtime' panicked at 'assertion failed: `(left == right)`
  left: `"test-birthtime-1\ntest-birthtime-2\n"`,
 right: `"test-birthtime-2\ntest-birthtime-1\n"`', /project/tests/by-util/test_ls.rs:496:5
```

**Review by sylvestre — CHANGES_REQUESTED** [review]
testsuite isn't green

**Comment by tertsdiepraam** [comment]
Adding another `sleep` between the first two `touch` calls might help with those tests.

**Comment by LevitatingBusinessMan** [comment]
Ye I know I kind of fucked that test up, I just haven't had the time to work on it again.

**Comment by LevitatingBusinessMan** [comment]
@tertsdiepraam 

I messed around a bit more with the sorting. It was hard for me to get that birthtime test to work because for some reason my machine gives a me perfectly reliable order of files, even when the timestamps are identical. Rust just gives the files in a specific order, every time. This isn't alphabetical or anything, it seems arbitrary, but it is still somehow related to the names of the files. It isn't the same order in other machines. I'll have to dig into Rust to figure out what causes that. But it's not relevant.

So if the timestamps are identical to files, we have them sorted in the order we discovered them. Either how Rust returned them to us, or in what order they showed up on the command line.

GNU's ls however, will always sort alphabetically as a fallback. So this is something we should definitely start doing as well.

Anyway the birthime test should just work like this:
```
touch x
sleep 1 ms
touch y
touch x
```
And the order should be `y x` because `y` was created later than `x`.

I think this works everywhere now, except on Windows where some of my tests have been timing out....

I'll commit the slight change to this branch now though.

**Comment by tertsdiepraam** [comment]
@LevitatingBusinessMan Those timestamps are definitely tricky! I ran into similar issues with the access time and I'd be very interested to hear if you find out what the causes are. The fallback to alphabetic ordering is a great idea, but I think we can just implement that in another PR. 

There are some conflicts in the `Cargo.toml` and `Cargo.lock`, could you fix those? That should start the CI too and then we can see whether the change worked. Thanks!

**Comment by LevitatingBusinessMan** [comment]
I am going mad

**Comment by tertsdiepraam** [comment]
@LevitatingBusinessMan Those build errors aren't really your fault. The `CmdResult` type was changed while you were working on this (see here for more info: https://github.com/uutils/coreutils/issues/1982). You can just replace all `result.stdout` with `result.stdout_str()` and it should be fine!

**Comment by sylvestre** [comment]
Once the conflict is fixed and the CI is green I will make sure it lands :)

**Comment by LevitatingBusinessMan** [comment]
Did you have to push changes to ls before I could fix the tests... :/

I was away on a sailing trip I couldn't fix them yesterday.
Now I got to resolve merge issues again.

**Comment by tertsdiepraam** [comment]
There are only some conflicts with the imports. The rest should merge without any issues!

**Comment by LevitatingBusinessMan** [comment]
The whole list of imports seem illogical to me anyway since the refactor. Why is so much under a `unix` condition? Anyway I'll try to merge it without any issues.

**Comment by tertsdiepraam** [comment]
There were the same number of platform-specific imports before, they were just spread out over multiple `use` statements with `cfg(unix)`. Here's why they're unix only:
* `HashMap` is used as cache for uid and gid, which don't exist on Windows
* `FileTypeExt` and `MetaDataExt` are platform-specific by nature
* `Duration` is used for modification time which also does not exist on Windows

**Comment by LevitatingBusinessMan** [comment]
@tertsdiepraam Seems like the push @sylvestre made while I was awaiting my test results broke some more stuff in the lockfile. Great

@tertsdiepraam In my opinion `HashMap` and `Duration` just shouldn't be imported like that, you can use them in full like `std::time::Duration` where necessary, which in both cases is only in a single location, makes the import list a lot less cluttered.

Anyhow I am again going to make a merge and if it builds on all platforms I'll push it to this branch.

**Comment by LevitatingBusinessMan** [comment]
Windows is failing, pushing to this branch regardless

**Comment by tertsdiepraam** [comment]
@LevitatingBusinessMan Well, let's agree to disagree😄 There is indeed another unix-specific import, which is not from `std` so I could merge that one with the imports from `std`. Personally, I like the merge-as-much-possible import style, but I can see why others might not like it.

**Comment by tertsdiepraam** [comment]
Aw dang, looks like `indoc` requirest Rust 1.45, maybe you could use `unindent` directly? That doesn't seem to have the same restriction. About the Windows test, I think we should be a bit lenient, because timestamps are kinda horrible. In one of the tests for access time I did this:
```rust
let file3_access = at.open("test-3").metadata().unwrap().accessed().unwrap();
let file4_access = at.open("test-4").metadata().unwrap().accessed().unwrap();

// It seems to be dependent on the platform whether the access time is actually set
if file3_access > file4_access {
    if cfg!(not(windows)) {
        result.stdout_only("test-3\ntest-4\ntest-2\ntest-1\n");
    } else {
        result.stdout_only("test-3  test-4  test-2  test-1\n");
    }
 } else {
    // Access time does not seem to be set on Windows and some other
    // systems so the order is 4 3 2 1
    if cfg!(not(windows)) {
        result.stdout_only("test-4\ntest-3\ntest-2\ntest-1\n");
    } else {
        result.stdout_only("test-4  test-3  test-2  test-1\n");
    }
}
```
It's kinda cheating but at least it gets tested properly for the platforms on which it works. I think you could do something similar for these tests and just not worry about it anymore.

**Comment by tertsdiepraam** [comment]
This explains a lot of our issues with Windows times:
> Not all file systems can record creation and last access time and not all file systems record them in the same manner. For example, on NT FAT, create time has a resolution of 10 milliseconds, write time has a resolution of 2 seconds, and access time has a resolution of 1 day (really, the access date). On NTFS, access time has a resolution of 1 hour. Therefore, the GetFileTime function may not return the same file time information set using the SetFileTime function. Furthermore, FAT records times on disk in local time. However, NTFS records times on disk in UTC. For more information, see File Times.

Source: https://docs.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-filetime

This might also explain the issue:
> A FILETIME structure that specifies when a file or directory is created. If the underlying file system does not support creation time, this member is zero (0).

So, maybe you could just only run the test if the creation time is not 0 (which should be Januari 1, 1601 according to the docs).

**Comment by LevitatingBusinessMan** [comment]
@tertsdiepraam, thanks for the research. I'll go ahead and look into it tomorrow. Hopefully I can close this then, school exams are coming up and I don't have that much spare time anymore.

**Comment by LevitatingBusinessMan** [comment]
@tertsdiepraam I could just increase the timeout in the test to be above 10ms (possibly, just in Windows builds), which should help with NTFS. There's no point in adding support for the tests for different filesystems because we don't run the tests in different filesystems, so we shouldn't have to expect the creation time might be zero.

Anyhow if I can't get it to succeed I am probably just going to remove this test for windows builds.

**Comment by LevitatingBusinessMan** [comment]
> Aw dang, looks like `indoc` requirest Rust 1.45

Where did you read this? I can use unindent directly but it's a bit more of a hassle.

**Comment by tertsdiepraam** [comment]
@LevitatingBusinessMan Sounds fine regarding the Windows tests. The rust version for indoc is on the GitHub page:

> Compiler requirement: rustc 1.45 or greater.

Source: https://github.com/dtolnay/indoc

**Comment by LevitatingBusinessMan** [comment]
@tertsdiepraam Seems like indoc uses `proc_macro` which isn't defined in Rust 1.40. I'll switch to unindent, too bad, because then the unindenting gets done on run-time.

I increased the timeout on windows and it failed: https://github.com/LevitatingBusinessMan/coreutils/runs/2451783244

**Comment by LevitatingBusinessMan** [comment]
@tertsdiepraam I cannot use `unindent`, Clap REFUSES to accept any string that isn't a static string literal it seems. I looked over it with someone and it seems rather impossible. I just added the `AFTER_HELP` string as a static string to the beginning of the file. There it is on root level anyway, so there's no need to use `indoc`.

Builds are succeeding on windows with the birthtime test turend off, it'll have to do for now. There is more I have to do anyway, like the `posix` stuff and the `FORMAT`. Stuff for after my exams.

**Comment by LevitatingBusinessMan** [comment]
@sylvestre This should definitely succeed.

**Comment by tertsdiepraam** [comment]
@LevitatingBusinessMan Why not do something like the literals here: https://stackoverflow.com/a/29483453. That doesn't require any additional packages.

**Comment by LevitatingBusinessMan** [comment]
@tertsdiepraam Raw strings don't solve the indentation issue.

I generally don't like using the backslash technique, because in some cases it makes the strings hard to read _and_ write.
Indoc exists just to resolve these issues.

But in this case using backslashes doesn't look too bad:

```RUST
        .after_help("\
            The TIME_STYLE argument can be full-iso, long-iso, iso.\n\
            Also the TIME_STYLE environment variable sets the default style to use.\n\
        ");
   ```

**Comment by sylvestre** [comment]
Sorry but some tests are failing:
```

---- test_ls::test_ls_order_birthtime stdout ----
current_directory_resolved: 
touch: /tmp/.tmpyINhjx/test-birthtime-1
touch: /tmp/.tmpyINhjx/test-birthtime-2
touch: /tmp/.tmpyINhjx/test-birthtime-1
run: /home/runner/work/coreutils/coreutils/target/debug/coreutils ls --time=birth -t
---- test_ls::test_ls_order_birthtime stderr ----
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"test-birthtime-1\ntest-birthtime-2\n"`,
 right: `"test-birthtime-2\ntest-birthtime-1\n"`', /home/runner/work/coreutils/coreutils/tests/by-util/test_ls.rs:576:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    test_ls::test_ls_order_birthtime
```

**Comment by LevitatingBusinessMan** [comment]
I might know what caused this.
I might this quick script:
```RUST
fn main() {
	let foo = std::fs::File::create("foo").unwrap();
	std::thread::sleep(std::time::Duration::from_millis(1));
	let bar = std::fs::File::create("bar").unwrap();
	if foo.metadata().unwrap().created().unwrap() != bar.metadata().unwrap().created().unwrap() {
		std::process::exit(0)
	}
	std::process::exit(1)
}
```
This binary will succeed on my SSD, but wil fail on my HDD. Why?
Linux buffers IO operations as to not too stress your disk too much, my theory is that this causes the files to be created at the same time.

I added a `sync_all` call to the first IO operation to make it happen immediately, and the problem was solved.
```RUST
fn main() {
	let foo = std::fs::File::create("foo").unwrap();
	foo.sync_all().unwrap();
	std::thread::sleep(std::time::Duration::from_millis(1));
	let bar = std::fs::File::create("bar").unwrap();
	if foo.metadata().unwrap().created().unwrap() != bar.metadata().unwrap().created().unwrap() {
		std::process::exit(0)
	}
	std::process::exit(1)
}
```
I was even able to remove the timeout, because the syncing of the buffer seems to take a few milliseconds. Anyhow I'll keep it in for the tests.

File: birth_test.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
+expected = <<-'END'
```
Comment (tertsdiepraam) [inline birth_test.rb:1]: I think this file was committed by accident?

File: tests/by-util/test_ls.rs [hunk 12]
```diff
@@ -558,6 +558,108 @@ fn test_ls_long_ctime() {
...
 }
 
+#[test]
+#[cfg(not(windows))]
+// This test is currently failing on windows
+fn test_ls_order_birthtime() {
+    let scene = TestScenario::new(util_name!());
+    let at = &scene.fixtures;
+
+    at.make_file("test-birthtime-1").sync_all().unwrap();
```
Comment (tertsdiepraam) [inline tests/by-util/test_ls.rs:568]: I think a small comment explaining why `sync_all` is there might be helpful.

**Review by tertsdiepraam — COMMENTED** [review]
Awesome! It looks like that fixed it. I'll probably do something similar for the access time once this is merged. Thanks so much for sticking with this PR!

File: birth_test.rb (hunk 11 above)
Comment (LevitatingBusinessMan) [reply birth_test.rb:1]: Definitely was, I'll just force push it out of the commit.

File: tests/by-util/test_ls.rs (hunk 12 above)
Comment (LevitatingBusinessMan) [reply tests/by-util/test_ls.rs:568]: Will do

File: tests/by-util/test_ls.rs [hunk 13]
```diff
@@ -565,6 +565,16 @@ fn test_ls_order_birthtime() {
...
+    /*
+        Here we make 2 files with a timeout in between.
+        After creating the first file try to sync it.
+        This ensures the file gets created immediately instead of being saved
+        inside the OS's IO operation buffer.
+        Without this, both files might accidentally be created at the same time,
+        even though we placed a timeout between creating the two.
+
+        https://github.com/uutils/coreutils/pull/1986/#issuecomment-828490651
+    */
```
Comment (LevitatingBusinessMan) [inline tests/by-util/test_ls.rs:577]: @tertsdiepraam This should suffice yes?

Comment (tertsdiepraam) [reply tests/by-util/test_ls.rs:577]: Definitely!

**Comment by tertsdiepraam** [comment]
I think removing it with another commit is fine!

**Review by tertsdiepraam — APPROVED** [review]

**Comment by sylvestre** [comment]
well done :)
